ui/src/features/stage/promotions.tsx

```diff
@@ -1,27 +1,80 @@
+import { createPromiseClient } from '@bufbuild/connect';
 import {
   faCircleCheck,
   faCircleExclamation,
   faCircleNotch,
   faCircleQuestion
 } from '@fortawesome/free-solid-svg-icons';
 import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
-import { useQuery } from '@tanstack/react-query';
+import { useQuery, useQueryClient } from '@tanstack/react-query';
 import { Popover, Table, Tooltip, theme } from 'antd';
 import { ColumnsType } from 'antd/es/table';
 import { format } from 'date-fns';
-import React from 'react';
+import React, { useEffect } from 'react';
 import { useParams } from 'react-router-dom';
 
+import { transport } from '@ui/config/transport';
 import { listPromotions } from '@ui/gen/service/v1alpha1/service-KargoService_connectquery';
+import { KargoService } from '@ui/gen/service/v1alpha1/service_connect';
+import { ListPromotionsResponse } from '@ui/gen/service/v1alpha1/service_pb';
 import { Promotion } from '@ui/gen/v1alpha1/types_pb';
 
 export const Promotions = () => {
+  const client = useQueryClient();
   const { name: projectName, stageName } = useParams();
   const { data: promotionsResponse, isLoading } = useQuery({
     ...listPromotions.useQuery({ project: projectName, stage: stageName }),
     enabled: !!stageName
   });
 
+  useEffect(() => {
+    if (isLoading || !promotionsResponse) {
```

Comment:
Is `loading` necessary here? If `promotionsResponse` is `undefined`, `isLoading` is true.